common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/AuthorizationResult.java

```diff
@@ -22,68 +22,50 @@
 // THE SOFTWARE.
 package com.microsoft.identity.common.internal.providers.oauth2;
 
-import android.content.Intent;
-
-import com.microsoft.identity.common.adal.internal.AuthenticationConstants;
-
 /**
  * A class to return the result of the authorization request to the calling code (ADAL or MSAL Controller classes)
  * This class should have a generic status in terms of : Cancelled, TimedOut, Error,  etc...
  * this class should also contain the AuthorizationResponse which contains the details returned from the
  * In the case of an error/exception this class should return the associated exception
  */
 public abstract class AuthorizationResult {
```

Comment:
Here's another place we can use the generic type parameter to prevent users of this class from casting their results.

```java
public abstract class AuthorizationResult
        <GenericAuthorizationResponse extends AuthorizationResponse,
         GenericAuthorizationErrorResponse extends AuthorizationErrorResponse> {

    // Other fields omitted for brevity...

    private GenericAuthorizationResponse mAuthorizationResponse;
    private GenericAuthorizationErrorResponse mAuthorizationErrorResponse;

    // Other accesssors omitted for brevity...

    public GenericAuthorizationResponse getAuthorizationResponse() {
        return mAuthorizationResponse;
    }

    public GenericAuthorizationErrorResponse getAuthorizationErrorResponse() {
        return mAuthorizationErrorResponse;
    }

    protected void setAuthorizationErrorResponse(final GenericAuthorizationErrorResponse authErrorResponse) {
        mAuthorizationErrorResponse = authErrorResponse;
    }

    protected void setAuthorizationResponse(final GenericAuthorizationResponse authResponse) {
        mAuthorizationResponse = authResponse;
    }
}
```

Subclasses like `MicrosoftAuthorizationResult` can then apply this pattern to more specific subclasses:

>Note: I would consider making `MicrosoftAuthorizationResult` `abstract`

```java
public class MicrosoftAuthorizationResult
        <GenericMicrosoftAuthorizationResponse extends MicrosoftAuthorizationResponse,
                GenericMicrosoftAuthorizationErrorResponse extends MicrosoftAuthorizationErrorResponse>
        extends AuthorizationResult<MicrosoftAuthorizationResponse, MicrosoftAuthorizationErrorResponse> {

    public MicrosoftAuthorizationResult(final AuthorizationStatus authStatus, final GenericMicrosoftAuthorizationResponse authResponse) {
        setAuthorizationStatus(authStatus);
        setAuthorizationResponse(authResponse);
    }

    public MicrosoftAuthorizationResult(final AuthorizationStatus authStatus, final GenericMicrosoftAuthorizationErrorResponse errorResponse) {
        setAuthorizationStatus(authStatus);
        setAuthorizationErrorResponse(errorResponse);
    }

}
```

Same idea for `AzureActiveDirectoryAuthorizationResult`

```java
public class AzureActiveDirectoryAuthorizationResult
        extends MicrosoftAuthorizationResult<AzureActiveDirectoryAuthorizationResponse, AzureActiveDirectoryAuthorizationErrorResponse>
```

And for `MicrosoftStsAuthorizationResult`

```java
public class MicrosoftStsAuthorizationResult
        extends MicrosoftAuthorizationResult<MicrosoftStsAuthorizationResponse, MicrosoftStsAuthorizationErrorResponse>
```

By doing this, subclasses will automatically enforce that all parameters are the proper-type. Users of these classes will never have to cast or `instanceof` their results, since the return-type is made explicit through the template

Reply:
I've left out the details of the accessor methods to save space/typing